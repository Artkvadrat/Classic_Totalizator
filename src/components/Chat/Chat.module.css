.button-delete {
    background-color: transparent;
    outline: none;
    border: none;
    padding: 0;
    margin: 0;
    cursor: pointer;
    font-weight: 500;
}
.button-delete:hover {
    color: red;
}