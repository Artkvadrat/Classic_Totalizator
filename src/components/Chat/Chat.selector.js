import { createSelector, createStructuredSelector } from 'reselect';
import parseDate from '../../services/dateParse/dateParse';

const tempArray = [
  {
    id: '3fa85f64-5717-b3fc-2c9',
    text: 'string',
    username: 'string',
    account_Id: '3fa85f64-5717-4562-b3fc-2c963f66afa6',
    avatarLink: 'https://avatars.dicebear.com/api/human/put-id-here.png',
    time: '2021-05-25T15:14:28.453Z'
  },
  {
    id: '3fa85f64-b3fc-2c9',
    text: 'string',
    username: 'string',
    account_Id: '3fa85f64-5717-4562-b3fc-2c963f66afa6',
    avatarLink: 'https://avatars.dicebear.com/api/human/put-id-here.png',
    time: '2021-05-25T15:14:28.453Z'
  },
  {
    id: '3fa85f64-5717-c9',
    text: 'string',
    username: 'string',
    account_Id: '3fa85f64-5717-4562-b3fc-2c963f66afa6',
    avatarLink: 'https://avatars.dicebear.com/api/human/put-id-here.png',
    time: '2021-05-25T15:14:28.453Z'
  },
  {
    id: '3fa85f64-5717-42c9',
    text: 'string',
    username: 'string',
    account_Id: '3fa85f64-5717-4562-b3fc-2c963f66afa6',
    avatarLink: 'https://avatars.dicebear.com/api/human/put-id-here.png',
    time: '2021-05-25T15:14:28.453Z'
  },
  {
    id: '3fa8-5717-4f66afa6',
    text: 'string',
    username: 'string',
    account_Id: '3fa85f64-5717-4562-b3fc-2c963f66afa6',
    avatarLink: 'https://avatars.dicebear.com/api/human/put-id-here.png',
    time: '2021-04-25T15:14:28.453Z'
  }
];

const dataSelector = (state) => state.chat;

const isLoadingSelector = createSelector(
  dataSelector,
  ({ isLoading }) => isLoading
);

const messagesSelector = createSelector(dataSelector, ({ messages }) =>
  tempArray
    .sort((a, b) => Date.parse(a.time) - Date.parse(b.time))
    .map(({ id, time, text, username, avatarLink }) => ({
      key: id,
      date: parseDate(time),
      message: text,
      userName: username,
      avatarImg: avatarLink
    }))
);

const chatSelector = createStructuredSelector({
  isLoading: isLoadingSelector,
  messages: messagesSelector
});

export default chatSelector;
