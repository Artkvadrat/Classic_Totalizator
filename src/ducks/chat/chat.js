/* eslint no-unneeded-ternary: ["error", { "defaultAssignment": true }] */
import HTTPService from '../../services/HTTPService/HTTPService';
import parseDate from '../../services/dateParse/dateParse';

export const REQUESTED = 'chat/requested';
export const RECEIVED = 'chat/received';
export const CREATED = 'event/create';
export const DELETED = 'event/delete';

const requested = () => ({
  type: REQUESTED
});

const received = (data) => ({
  type: RECEIVED,
  payload: data
});

const created = () => ({
  type: CREATED
});

const deleted = () => ({
  type: DELETED
});

export const deleteMessage = (id) => (dispatch) =>
  HTTPService.request({
    method: 'DELETE',
    path: `/api/Chat/${id}`
  }).then(dispatch(deleted()));

export const getMessages = () => (dispatch) => {
  dispatch(requested());

  return HTTPService.request({ path: '/api/Chat' }).then((data) => {
    const array = data.messages
      .sort((a, b) => Date.parse(a.time) - Date.parse(b.time))
      .map(({ id, time, text, username, avatarLink }) => ({
        key: id,
        date: parseDate(time),
        message: text,
        userName: username,
        avatarImg: avatarLink
          ? avatarLink
          : 'https://avatars.dicebear.com/api/bottts/.png'
      }));
    dispatch(received(array));
  });
};

export const createMessage = (data) => (dispatch) =>
  HTTPService.request({
    method: 'POST',
    path: '/api/Chat',
    body: { text: data }
  }).then(dispatch(created()));

const initialState = {
  isLoading: false,
  messages: []
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case REQUESTED:
      return {
        ...state,
        isLoading: true
      };

    case RECEIVED:
      return {
        ...state,
        isLoading: false,
        messages: [...action.payload]
      };

    case CREATED:
      return {
        ...initialState
      };
    case DELETED:
      return {
        ...initialState
      };
    default:
      return state;
  }
};

export default reducer;
