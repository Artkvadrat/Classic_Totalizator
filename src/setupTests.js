import enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-17-updated';

enzyme.configure({ adapter: new Adapter() });

const localStorageMock = {
  getItem: jest.fn(),
  setItem: jest.fn(),
  removeItem: jest.fn(),
  clear: jest.fn()
};
window.localStorage = localStorageMock;
