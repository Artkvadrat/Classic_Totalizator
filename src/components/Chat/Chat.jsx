import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Skeleton, Table, Row, Col, Space } from 'antd';
import CreatorMessages from '../CreatorMessages/CreatorMessages';
import { getMessages, deleteMessage } from '../../ducks/chat/chat';
import useInterval from '../../services/useInterval/useInterval';
import styles from './Chat.module.css';

const Chat = () => {
  const { isLoading, messages } = useSelector((state) => state.chat);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getMessages());
  }, [dispatch]);

  useInterval(() => {
    dispatch(getMessages());
  }, 5000);

  const handleClick = (event) => {
    dispatch(deleteMessage(event.target.value)).then(() =>
      dispatch(getMessages())
    );
  };

  return (
    <>
      <Row justify="space-between" align="middle">
        <Col>
          <h1>All messages</h1>
        </Col>
        <Col style={{ margin: 16 }}>
          <CreatorMessages />
        </Col>
      </Row>

      {isLoading ? (
        <Skeleton active />
      ) : (
        <Table dataSource={messages}>
          <Table.Column
            title="User"
            key="user"
            render={({ userName, avatarImg }) => (
              <Space size="small">
                <img height="30px" src={avatarImg} alt="avatar" />
                <span>{userName}</span>
              </Space>
            )}
          />
          <Table.Column title="Message" dataIndex="message" key="message" />
          <Table.Column title="Time" dataIndex="date" key="date" />
          <Table.Column
            title="Action"
            key="action"
            render={({ key }) => (
              <button
                className={styles['button-delete']}
                value={key}
                onClick={handleClick}
                type="button"
              >
                Delete
              </button>
            )}
          />
        </Table>
      )}
    </>
  );
};

export default Chat;
