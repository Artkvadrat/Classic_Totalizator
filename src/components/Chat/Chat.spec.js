import React from 'react';
import { shallow, mount } from 'enzyme';
import { Skeleton, Table } from 'antd';
import { useDispatch, useSelector } from 'react-redux';
import { getMessages, deleteMessage } from '../../ducks/chat/chat';

import '../../setupTests';

import Chat from './Chat';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn()
}));

jest.mock('../../ducks/chat/chat', () => ({
  deleteMessage: jest.fn(),
  getMessages: jest.fn()
}));

describe('CreatorMessages component', () => {
  const dispatch = jest.fn(() => Promise.resolve());
  beforeAll(() => {
    window.matchMedia =
      window.matchMedia ||
      function () {
        return {
          matches: false,
          addListener: () => {},
          removeListener: () => {}
        };
      };
  });

  beforeEach(() => {
    useDispatch.mockReturnValue(dispatch);
    useSelector.mockReturnValue({
      isLoading: false,
      messages: [
        {
          key: '3fa85f64-5717-4562-b3fc-2c963f66afa6',
          message: 'string',
          userName: 'Vova',
          avatarImg: 'https://avatars.dicebear.com/api/bottts/.png',
          date: '18:20 05.04.2021'
        }
      ]
    });
    deleteMessage.mockReturnValue(Symbol.for('delete'));
    getMessages.mockReturnValue(Symbol.for('get'));
    jest.clearAllMocks();
  });
  it('should dispatch getMessages on mount', () => {
    expect(dispatch).not.toHaveBeenCalled();
    expect(getMessages).not.toHaveBeenCalled();
    mount(<Chat />);
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith(Symbol.for('get'));
    expect(getMessages).toHaveBeenCalledTimes(1);
  });
  it('should render creator messages and Skeleton', () => {
    useSelector.mockReturnValue({
      isLoading: true,
      messages: [
        {
          key: '3fa85f64-5717-4562-b3fc-2c963f66afa6',
          message: 'string',
          userName: 'Vova',
          avatarImg: 'https://avatars.dicebear.com/api/bottts/.png',
          date: '18:20 05.04.2021'
        }
      ]
    });
    const result = shallow(<Chat />);
    const creator = result.find(`CreatorMessages`);
    const skeleton = result.find(Skeleton);

    expect(creator.exists()).toEqual(true);
    expect(skeleton.length).toEqual(1);
    expect(skeleton.prop('active')).toEqual(true);

    expect(result).toMatchSnapshot();
  });
  it('should render creator messages and messages', () => {
    const result = shallow(<Chat />);
    const creator = result.find(`CreatorMessages`);
    const table = result.find(Table);

    expect(table.exists()).toEqual(true);
    expect(creator.exists()).toEqual(true);

    expect(result).toMatchSnapshot();
  });
  it('should delete message on click', () => {
    const result = mount(<Chat />);
    const fakeEvent = {
      target: {
        value: 'dsasdascads'
      }
    };

    const buttonDelete = result.find('.button-delete');
    buttonDelete.simulate('click', fakeEvent);

    expect(dispatch).toHaveBeenCalledTimes(2);
    expect(deleteMessage).toHaveBeenCalledTimes(1);
    expect(deleteMessage).toHaveBeenCalledWith(fakeEvent.target.value);
    expect(getMessages).toHaveBeenCalledTimes(1);
  });
});
