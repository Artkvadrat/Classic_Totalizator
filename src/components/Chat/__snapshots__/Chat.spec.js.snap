// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatorMessages component should render creator messages and Skeleton 1`] = `
<Fragment>
  <Row
    align="middle"
    justify="space-between"
  >
    <Col>
      <h1>
        All messages
      </h1>
    </Col>
    <Col
      style={
        Object {
          "margin": 16,
        }
      }
    >
      <CreatorMessages />
    </Col>
  </Row>
  <Skeleton
    active={true}
    avatar={false}
    paragraph={true}
    title={true}
  />
</Fragment>
`;

exports[`CreatorMessages component should render creator messages and messages 1`] = `
<Fragment>
  <Row
    align="middle"
    justify="space-between"
  >
    <Col>
      <h1>
        All messages
      </h1>
    </Col>
    <Col
      style={
        Object {
          "margin": 16,
        }
      }
    >
      <CreatorMessages />
    </Col>
  </Row>
  <Table
    dataSource={
      Array [
        Object {
          "avatarImg": "https://avatars.dicebear.com/api/bottts/.png",
          "date": "18:20 05.04.2021",
          "key": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
          "message": "string",
          "userName": "Vova",
        },
      ]
    }
    rowKey="key"
  >
    <Column
      key="user"
      render={[Function]}
      title="User"
    />
    <Column
      dataIndex="message"
      key="message"
      title="Message"
    />
    <Column
      dataIndex="date"
      key="date"
      title="Time"
    />
    <Column
      key="action"
      render={[Function]}
      title="Action"
    />
  </Table>
</Fragment>
`;
